#Machine_Learning_Project2(Placement Prediction )

* **Purpose**:To analyse the Campus Recruitment data and predict the placements. 

The input Dataset has the following columns:

* sl_no : Serial Number
* gender :'M' & 'F'
* ssc_p : Secondary Education percentage
* ssc_b : Board of Education(ssc)
* hsc_p : Higher Secondary Education percentage
* hsc_b : Board of Education(hsc)
* hsc_s : Specialization in Higher Secondary Education
* degree_p : Degree Percentage
* degree_t : Field of degree education
* workex : Work Experience
* etest_p : Employability test percentage
* specialisation : Specialization in MBA
* mba_p : MBA percentage
* status : Status of placement
* salary : Salary offered to candidates

#Importing Libraries:

In [1]:
import numpy as np
import pandas as pd
import matplotlib as plt

#Reading data:

In [2]:
pdataset = pd.read_csv('/content/Placement_Data_Full_Class.csv')

In [3]:
pdataset

sl_no gender  ssc_p    ssc_b  ...  specialisation  mba_p      status    salary
0        1      M  67.00   Others  ...          Mkt&HR  58.80      Placed  270000.0
1        2      M  79.33  Central  ...         Mkt&Fin  66.28      Placed  200000.0
2        3      M  65.00  Central  ...         Mkt&Fin  57.80      Placed  250000.0
3        4      M  56.00  Central  ...          Mkt&HR  59.43  Not Placed       NaN
4        5      M  85.80  Central  ...         Mkt&Fin  55.50      Placed  425000.0
..     ...    ...    ...      ...  ...             ...    ...         ...       ...
210    211      M  80.60   Others  ...         Mkt&Fin  74.49      Placed  400000.0
211    212      M  58.00   Others  ...         Mkt&Fin  53.62      Placed  275000.0
212    213      M  67.00   Others  ...         Mkt&Fin  69.72      Placed  295000.0
213    214      F  74.00   Others  ...          Mkt&HR  60.23      Placed  204000.0
214    215      M  62.00  Central  ...          Mkt&HR  60.22  Not Placed       NaN

[215 rows x 15 columns]

In [4]:
pdataset.head()

sl_no gender  ssc_p    ssc_b  ...  specialisation  mba_p      status    salary
0      1      M  67.00   Others  ...          Mkt&HR  58.80      Placed  270000.0
1      2      M  79.33  Central  ...         Mkt&Fin  66.28      Placed  200000.0
2      3      M  65.00  Central  ...         Mkt&Fin  57.80      Placed  250000.0
3      4      M  56.00  Central  ...          Mkt&HR  59.43  Not Placed       NaN
4      5      M  85.80  Central  ...         Mkt&Fin  55.50      Placed  425000.0

[5 rows x 15 columns]

In [5]:
pdataset.tail()

sl_no gender  ssc_p    ssc_b  ...  specialisation  mba_p      status    salary
210    211      M   80.6   Others  ...         Mkt&Fin  74.49      Placed  400000.0
211    212      M   58.0   Others  ...         Mkt&Fin  53.62      Placed  275000.0
212    213      M   67.0   Others  ...         Mkt&Fin  69.72      Placed  295000.0
213    214      F   74.0   Others  ...          Mkt&HR  60.23      Placed  204000.0
214    215      M   62.0  Central  ...          Mkt&HR  60.22  Not Placed       NaN

[5 rows x 15 columns]

In [6]:
pdataset.describe()

sl_no       ssc_p  ...       mba_p         salary
count  215.000000  215.000000  ...  215.000000     148.000000
mean   108.000000   67.303395  ...   62.278186  288655.405405
std     62.209324   10.827205  ...    5.833385   93457.452420
min      1.000000   40.890000  ...   51.210000  200000.000000
25%     54.500000   60.600000  ...   57.945000  240000.000000
50%    108.000000   67.000000  ...   62.000000  265000.000000
75%    161.500000   75.700000  ...   66.255000  300000.000000
max    215.000000   89.400000  ...   77.890000  940000.000000

[8 rows x 7 columns]

In [7]:
#checking for null values
pdataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 215 entries, 0 to 214
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sl_no           215 non-null    int64  
 1   gender          215 non-null    object 
 2   ssc_p           215 non-null    float64
 3   ssc_b           215 non-null    object 
 4   hsc_p           215 non-null    float64
 5   hsc_b           215 non-null    object 
 6   hsc_s           215 non-null    object 
 7   degree_p        215 non-null    float64
 8   degree_t        215 non-null    object 
 9   workex          215 non-null    object 
 10  etest_p         215 non-null    float64
 11  specialisation  215 non-null    object 
 12  mba_p           215 non-null    float64
 13  status          215 non-null    object 
 14  salary          148 non-null    float64
dtypes: float64(6), int64(1), object(8)
memory usage: 25.3+ KB


#Analysing Data:


* Check the values in columns i.e. categorical values using  value_counts() function

In [8]:
pdataset.gender.value_counts()

M    139
F     76
Name: gender, dtype: int64

In [9]:
pdataset.ssc_b.value_counts()

Central    116
Others      99
Name: ssc_b, dtype: int64

In [10]:
pdataset.hsc_b.value_counts()

Others     131
Central     84
Name: hsc_b, dtype: int64

In [11]:
pdataset.hsc_s.value_counts()

Commerce    113
Science      91
Arts         11
Name: hsc_s, dtype: int64

In [12]:
pdataset.degree_t.value_counts()

Comm&Mgmt    145
Sci&Tech      59
Others        11
Name: degree_t, dtype: int64

In [13]:
pdataset.workex.value_counts()

No     141
Yes     74
Name: workex, dtype: int64

In [14]:
pdataset.specialisation.value_counts()

Mkt&Fin    120
Mkt&HR      95
Name: specialisation, dtype: int64

In [15]:
pdataset.status.value_counts()

Placed        148
Not Placed     67
Name: status, dtype: int64

* hsc_s & degree_t are having 3 classes, whereas others hav just 2 classes.
* Also 148 students are placed, & 67 students are not placed.

In [16]:
#As data in the salary column is not available,we can drop the whole salary column.
pdataset.drop(columns=['salary'])


sl_no gender  ssc_p    ssc_b  ...  etest_p specialisation  mba_p      status
0        1      M  67.00   Others  ...     55.0         Mkt&HR  58.80      Placed
1        2      M  79.33  Central  ...     86.5        Mkt&Fin  66.28      Placed
2        3      M  65.00  Central  ...     75.0        Mkt&Fin  57.80      Placed
3        4      M  56.00  Central  ...     66.0         Mkt&HR  59.43  Not Placed
4        5      M  85.80  Central  ...     96.8        Mkt&Fin  55.50      Placed
..     ...    ...    ...      ...  ...      ...            ...    ...         ...
210    211      M  80.60   Others  ...     91.0        Mkt&Fin  74.49      Placed
211    212      M  58.00   Others  ...     74.0        Mkt&Fin  53.62      Placed
212    213      M  67.00   Others  ...     59.0        Mkt&Fin  69.72      Placed
213    214      F  74.00   Others  ...     70.0         Mkt&HR  60.23      Placed
214    215      M  62.00  Central  ...     89.0         Mkt&HR  60.22  Not Placed

[215 rows x 14 columns]

#Preprocessing data for performing algorithms:

* The attribute *status* will become our '**label**' and the rest of the attributes will become the '**features**'.

In [17]:
#separating the features and labels into two different dataframes.
x = pdataset.iloc[:,[4,7,9,10,11,12]].values
y = pdataset.iloc[:, -2].values

In [18]:
x

array([[91.0, 58.0, 'No', 55.0, 'Mkt&HR', 58.8],
       [78.33, 77.48, 'Yes', 86.5, 'Mkt&Fin', 66.28],
       [68.0, 64.0, 'No', 75.0, 'Mkt&Fin', 57.8],
       ...,
       [67.0, 73.0, 'Yes', 59.0, 'Mkt&Fin', 69.72],
       [66.0, 58.0, 'No', 70.0, 'Mkt&HR', 60.23],
       [58.0, 53.0, 'No', 89.0, 'Mkt&HR', 60.22]], dtype=object)

In [19]:
y

array(['Placed', 'Placed', 'Placed', 'Not Placed', 'Placed', 'Not Placed',
       'Not Placed', 'Placed', 'Placed', 'Not Placed', 'Placed', 'Placed',
       'Not Placed', 'Placed', 'Not Placed', 'Placed', 'Placed',
       'Not Placed', 'Not Placed', 'Placed', 'Placed', 'Placed', 'Placed',
       'Placed', 'Placed', 'Not Placed', 'Placed', 'Placed', 'Placed',
       'Not Placed', 'Placed', 'Not Placed', 'Placed', 'Placed',
       'Not Placed', 'Placed', 'Not Placed', 'Placed', 'Placed', 'Placed',
       'Placed', 'Not Placed', 'Not Placed', 'Placed', 'Placed',
       'Not Placed', 'Not Placed', 'Placed', 'Placed', 'Not Placed',
       'Placed', 'Not Placed', 'Not Placed', 'Placed', 'Placed', 'Placed',
       'Placed', 'Placed', 'Placed', 'Placed', 'Placed', 'Placed',
       'Placed', 'Not Placed', 'Placed', 'Not Placed', 'Placed', 'Placed',
       'Not Placed', 'Placed', 'Placed', 'Placed', 'Placed', 'Placed',
       'Placed', 'Not Placed', 'Placed', 'Placed', 'Placed', 'Not Placed',
  

##Encoding Textual information
* Label Encoder
* One Hot Encoder

In [20]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct=ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[2,4])],remainder='passthrough')

In [21]:
x=np.array(ct.fit_transform(x))

In [22]:
from sklearn.preprocessing import LabelEncoder
le_en=LabelEncoder()

In [23]:
y=le_en.fit_transform(y)

In [24]:
x

array([[1.0, 0.0, 0.0, ..., 58.0, 55.0, 58.8],
       [0.0, 1.0, 1.0, ..., 77.48, 86.5, 66.28],
       [1.0, 0.0, 1.0, ..., 64.0, 75.0, 57.8],
       ...,
       [0.0, 1.0, 1.0, ..., 73.0, 59.0, 69.72],
       [1.0, 0.0, 0.0, ..., 58.0, 70.0, 60.23],
       [1.0, 0.0, 0.0, ..., 53.0, 89.0, 60.22]], dtype=object)

In [25]:
y

array([1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1,
       1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0])

#Split into Test and Train data:

In [26]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y,test_size=0.3,random_state=0)


In [27]:
print(len(x_train))
print(len(x_test))
print(len(y_train))
print(len(y_test))

150
65
150
65


In [28]:
x_train

array([[1.0, 0.0, 0.0, ..., 61.0, 60.0, 60.64],
       [1.0, 0.0, 0.0, ..., 58.0, 62.0, 60.59],
       [1.0, 0.0, 1.0, ..., 78.86, 97.4, 74.01],
       ...,
       [1.0, 0.0, 1.0, ..., 73.0, 80.0, 67.05],
       [0.0, 1.0, 1.0, ..., 57.0, 78.0, 54.55],
       [1.0, 0.0, 0.0, ..., 56.0, 84.0, 52.64]], dtype=object)

In [29]:
y_train

array([0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1])

In [30]:
x_test

array([[1.0, 0.0, 0.0, 1.0, 70.0, 65.0, 88.0, 71.96],
       [1.0, 0.0, 0.0, 1.0, 76.0, 65.6, 58.0, 55.47],
       [0.0, 1.0, 0.0, 1.0, 75.0, 69.0, 62.0, 62.36],
       [0.0, 1.0, 0.0, 1.0, 51.0, 58.0, 60.0, 58.79],
       [0.0, 1.0, 1.0, 0.0, 80.0, 72.0, 63.79, 66.04],
       [1.0, 0.0, 0.0, 1.0, 62.0, 77.5, 74.0, 67.0],
       [0.0, 1.0, 1.0, 0.0, 70.0, 76.0, 66.0, 64.44],
       [1.0, 0.0, 0.0, 1.0, 63.0, 72.0, 56.0, 60.41],
       [0.0, 1.0, 1.0, 0.0, 49.8, 67.25, 55.0, 51.58],
       [0.0, 1.0, 1.0, 0.0, 79.0, 68.0, 84.0, 66.69],
       [1.0, 0.0, 0.0, 1.0, 66.6, 65.0, 71.0, 52.71],
       [1.0, 0.0, 0.0, 1.0, 65.66, 71.25, 72.0, 63.23],
       [0.0, 1.0, 0.0, 1.0, 74.0, 65.0, 75.0, 72.29],
       [1.0, 0.0, 0.0, 1.0, 61.0, 61.0, 58.0, 53.94],
       [1.0, 0.0, 1.0, 0.0, 64.8, 70.2, 84.27, 67.2],
       [1.0, 0.0, 0.0, 1.0, 61.33, 56.87, 66.0, 58.3],
       [0.0, 1.0, 1.0, 0.0, 61.0, 75.5, 70.0, 68.2],
       [1.0, 0.0, 0.0, 1.0, 55.0, 65.0, 62.0, 65.04],
       [0.0, 1.0, 1.0, 0.

In [31]:
y_test

array([0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1])

#Scaling_data:

###Standardising Data

In [32]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train[:,4:] = sc.fit_transform(x_train[:,4:] )
x_test[:,4:]  = sc.fit_transform(x_test[:,4:] )

#Applying different Algorithms:

##Comparing in Classification Models

* Logistic Regression
* KNN
* Decision Tree
* Random Forest 
* Linear SVC
* Kernel SVC

In [33]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC


In [34]:
l_cla = LogisticRegression()
k_cla = KNeighborsClassifier()
d_cla = DecisionTreeClassifier()
r_cla = RandomForestClassifier()
s_cla = SVC(kernel='linear')
ks_cla = SVC(kernel= 'rbf')

In [35]:
#Executing fit function on all
l_cla.fit(x_train, y_train)
k_cla.fit(x_train, y_train)
d_cla.fit(x_train, y_train)
r_cla.fit(x_train, y_train)
s_cla.fit(x_train, y_train)
ks_cla.fit(x_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [36]:
#Using Predict Function
l_pred = l_cla.predict(x_test)
k_pred = k_cla.predict(x_test)
d_pred = d_cla.predict(x_test)
r_pred = r_cla.predict(x_test)
s_pred = s_cla.predict(x_test)
ks_pred = ks_cla.predict(x_test)

In [37]:
#Using confusion_matrix Function
from sklearn.metrics import confusion_matrix

In [38]:
l_c = confusion_matrix(y_test, l_pred)
k_c = confusion_matrix(y_test, k_pred)
d_c = confusion_matrix(y_test, d_pred)
r_c = confusion_matrix(y_test, r_pred)
s_c = confusion_matrix(y_test, s_pred)
ks_c = confusion_matrix(y_test, ks_pred)

In [39]:
print(l_c)


[[13  6]
 [ 4 42]]


In [40]:
print(k_c)

[[11  8]
 [ 2 44]]


In [41]:
print(d_c)

[[ 9 10]
 [ 6 40]]


In [42]:
print(r_c)

[[13  6]
 [ 4 42]]


In [43]:
print(s_c)

[[12  7]
 [ 3 43]]


In [44]:
print(ks_c)

[[10  9]
 [ 2 44]]


In [45]:
#Using accuracy_score Function.
from sklearn.metrics import accuracy_score

In [46]:
l_a = accuracy_score(y_test, l_pred)
k_a = accuracy_score(y_test, k_pred)
d_a = accuracy_score(y_test, d_pred)
r_a = accuracy_score(y_test, r_pred)
s_a = accuracy_score(y_test, s_pred)
ks_a = accuracy_score(y_test, ks_pred)

In [47]:
print('Logistic Regression: ' + str(l_a) + '\nKNN: ' + str(k_a) + '\nDecision Tree: ' + str(d_a) +
      '\nRandom Forest: ' + str(r_a) + '\nLinear SVC: ' + str(s_a) + '\nKernel SVC: ' + str(ks_a))

Logistic Regression: 0.8461538461538461
KNN: 0.8461538461538461
Decision Tree: 0.7538461538461538
Random Forest: 0.8461538461538461
Linear SVC: 0.8461538461538461
Kernel SVC: 0.8307692307692308


We reach to the end of this prediction, concluding that:
* We obtain the highest accuracy of 84% by most of the algorithms.



*Thank You*

**Kharat Pooja C**